pkg/k8s/service_cache.go

```diff
@@ -274,6 +274,16 @@ func (s *ServiceCache) DeleteService(k8sSvc *slim_corev1.Service, swg *lock.Stop
 	}
 }
 
+// IsServiceLocal returns whether the given service is known, and all
+// backends belong to the local cluster.
+func (s *ServiceCache) IsServiceLocal(svcID ServiceID) bool {
+	s.mutex.RLock()
```

Comment:
Doesn't this lead to a dead lock?

1. `SyncWithK8sFinished` does `Service.Lock()`
2. `ServiceCache` gets a new service and sends it to `Events` channel
3. `watcher.K8sServiceHandler` reads an event from a channel and calls `svcManager.UpsertService`. Blocks acquiring `Service.Lock()`
4. `ServiceCache` gets another service and sends it to `Events` channel (repeat here until buffer is full)
5. `ServiceCache` gets another service, takes the `ServiceCache` write lock, and under the lock tries to send it to `Events` channel
6. `ServiceManager` calls `IsServiceLocal` and tries to acquire `ServiceCache` read lock.

We now have a deadlock with two goroutines:
- `ServiceManager.SyncWithK8sFinished`: `ServiceManager` is write locked and it tries to acquire `ServiceCache` read lock.
- `ServiceCache.UpdateService`: `ServiceCache` is write locked and it tries to indirectly lock `ServiceManager` due to a blocking send to a full channel.

Could you try to avoid this cyclic locking issue by passing the `IsServiceLocal` property down to `ServiceManager` as part of `lb.SVC`? Or alternatively you could pass a map of these properties to `SyncWithK8sFinished`.

Reply:
Or we can probably fix this by releasing the `ServiceCache` lock before the send to the `Events` channel. It probably would also be safer to drop the buffering on the `Events` channel since that really makes it harder to catch these sort of hidden lurking dead locks.

Reply:
Hmm, you are definitely right about the possible deadlock. Let me rework it to prevent this possibility.

Reply:
The list of local services is now retrieved in advance, as you suggested, so that we don't need to hold both locks at the same time while checking whether a given service is local or not.